### PR TITLE
ENH Add CUDA 11.2.2 CTK

### PR DESCRIPTION
**Pull request by mike-wendt**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

- Added CUDA 11.2.2 version to the repo as it was just released
- Kept build version at `8`

### CODE REVIEW

File: conda-forge.yml
Comment (kkraus14) [inline conda-forge.yml:4]: Does this need to be patched somewhere else? @jakirkham

Comment (mike-wendt) [reply conda-forge.yml:4]: Screen Shot 2021-03-16 at 14 25 58
Looks like a recent change for 11.2.1+

Comment (kkraus14) [reply conda-forge.yml:4]: Yup, I just vaguely remember that this file is autogenerated from somewhere else and that a re-render would blow away these changes. I could be mistaken though.

Comment (mike-wendt) [reply conda-forge.yml:4]: Yeah I need to rerender once more to get the change propagated and get the windows builds working

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:262]: We skip the build number bump since we are not changing the older version packages

File: conda-forge.yml
Comment (jakirkham) [reply conda-forge.yml:4]: I think it is just a re-render, but let's check with @jaimergp 🙂

**Comment by mike-wendt** [comment]
FYI I've been seeing a lot of failures with retrieving the CUDA runfiles and the site itself takes a while to load. May have to wait and re-run the CI later when the CDN is in a better state.

File: conda-forge.yml
Comment (jaimergp) [reply conda-forge.yml:4]: I think we need that to provide the drivers in case we want to run some tests, but without getting into a catch 22 situation if we use `conda-forge-ci-setup` provided installers. If that's correct, any sufficiently recent one should do, right? Maybe @scopatz better recalls why `choco` was needed.

**Comment by jakirkham** [comment]
Restarted the failed builds. They seem to be doing better now

Is there some way to see the CDN status for the site hosting the runfiles publicly?

**Comment by mike-wendt** [comment]
> Restarted the failed builds. They seem to be doing better now
> 
> Is there some way to see the CDN status for the site hosting the runfiles publicly?

I don't know of a status page off hand. I'll see if there is any, I was just experiencing long load times for the site itself and then small files like the md5 sums page

**Comment by mike-wendt** [comment]
@jakirkham looks like the builds succeeded now, thanks for re-running them

File: conda-forge.yml
Comment (jakirkham) [reply conda-forge.yml:4]: So what would be the next step here? Merge this and open a PR to `conda-forge-ci-setup`? Reverse order? Something else?

**Comment by mike-wendt** [comment]
FYI there is discussion in #56 about how `ppc64le` builds require CentOS 8 from CUDA 11.0+

File: conda-forge.yml
Comment (jakirkham) [reply conda-forge.yml:4]: ^ @jaimergp?

**Comment by kkraus14** [comment]
> FYI there is discussion in #56 about how `ppc64le` builds require CentOS 8 from CUDA 11.0+

Can we conditionally skip ppc64le for now in the problem cases to get this in?

**Comment by jakirkham** [comment]
For context, I'm trying to figure out what we need to here ( https://github.com/conda-forge/cudatoolkit-feedstock/pull/57#discussion_r596262798 )

**Comment by jakirkham** [comment]
@jaimergp would you be able to advise on next steps above ( https://github.com/conda-forge/cudatoolkit-feedstock/pull/57#discussion_r596262798 )? 🙂 Trying to understand what we need to do next

**Comment by leofang** [comment]
Didn't @jaimergp write some great docs for us to follow? 🙂
https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=cuda#adding-support-for-a-new-cuda-version

**Comment by jakirkham** [comment]
Right I'm trying to understand this comment ( https://github.com/conda-forge/cudatoolkit-feedstock/pull/57#discussion_r595556517 ) and what that means in terms of what we should do

**Comment by leofang** [comment]
Ah ok, I saw you linked to this (https://github.com/conda-forge/cudatoolkit-feedstock/pull/57#discussion_r596262798) and was wondering. Don't mind me.

**Comment by jaimergp** [comment]
Sorry, I am having some teaching these weeks and couldn't attend these messages. 

We are still in the 11.2 branch, right? That means that the "build-time" infrastructure should still work.

Ideally, though, we need a PR in `conda-forge-ci-setup` similar to https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/143 (add 11.2.2 and make it default for 11.2).

I am still unsure about that `choco` line. I guess it doesn't hurt to update it to use the latest one (specially since `choco` likes to stop supporting prior versions rather quickly), but at some point we should investigate whether that's still needed.

**Comment by jakirkham** [comment]
Submitted these PRs

https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/147
https://github.com/conda-forge/docker-images/pull/177

**Comment by jakirkham** [comment]
No worries Jaime! Thanks for the reply 😄

Yeah I wonder if it is not needed, but so seldom use Windows any more that I'm probably not the best person to advise there 😅

Do you know which order these go in? 🙂 I think we are still waiting on an upstream Docker image PR before that one can land

**Comment by jaimergp** [comment]
> Do you know which order these go in? 

They are technically uncoupled, but we need both to have this PR merged. I think you can merge `conda-forge-ci-setup` and then we wait for `docker-images` before tackling this one.

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

**Review by leofang — COMMENTED** [review]
Propagating a nice practice from other feedstocks. To fix the linter error, [we (Keith, John and I) verified](https://github.com/conda-forge/staged-recipes/pull/14276#discussion_r595514818) that this should work: Change line 318 as follows:
```
license: LicenseRef-NVIDIA-End-User-License-Agreement
```
@mike-wendt I know this is not in the scope of this PR, but could you help us make the change and silence the linter? 🙂

**Comment by jakirkham** [comment]
This is a good idea Leo. That said, can I ask that we save this for a future PR? 🙂

**Comment by leofang** [comment]
@conda-forge-admin, please rerender

**Review by leofang — APPROVED** [review]

File: recipe/conda_build_config.yaml
Comment (isuruf) [inline recipe/conda_build_config.yaml:9]: Do we need to keep supporting `11.2` and `11.2.1`? Isn't `11.2.2` enough?

**Comment by leofang** [comment]
> This is a good idea Leo. That said, can I ask that we save this for a future PR? 🙂

I took the liberty and pushed to @mike-wendt's branch. I didn't mean to block, but it doesn't deserve its own PR IMHO, sorry.

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render

File: recipe/conda_build_config.yaml
Comment (jaimergp) [reply recipe/conda_build_config.yaml:9]: AFAIK, only 12.2.2 will be used in the builds.

Comment (jakirkham) [reply recipe/conda_build_config.yaml:9]: Yeah this only affects the package builds in this feedstocks

**Comment by leofang** [comment]
@conda-forge/cudatoolkit CI is all green.

**Review by kkraus14 — APPROVED** [review]

File: recipe/conda_build_config.yaml
Comment (isuruf) [reply recipe/conda_build_config.yaml:9]: Sure, but what's the point. We'll have to remember to remove them if we bump the build number in a new PR. So, I think we should just remove building them

**Review by mike-wendt — COMMENTED** [review]
LGTM 👍

File: recipe/conda_build_config.yaml
Comment (jakirkham) [reply recipe/conda_build_config.yaml:9]: Why do we have to remember to remove them? If the package has a systematic bug (as has come up with EULA statement script before), we would want to fix older packages as well

Comment (isuruf) [reply recipe/conda_build_config.yaml:9]: Ok

Comment (isuruf) [reply recipe/conda_build_config.yaml:9]: On second thought, even if the build number is not bumped, the hash might change, leading to new packages being built which is a waste.

**Comment by jakirkham** [comment]
Thanks Mike for the PR and everyone for the reviews! 😄

Will work on getting Docker images and `conda-forge-ci-setup` out after these package are up

**Comment by jakirkham** [comment]
2 of the Windows builds got stuck trying to upload packages. Canceled them and restarted those builds

File: recipe/conda_build_config.yaml
Comment (jakirkham) [reply recipe/conda_build_config.yaml:9]: If that were the case, it would imply we should have bumped the build number, but didn't

In any event if we look at Anaconda.org, we can see there are only 2 packages uploaded recently for Linux (both 11.2.2). Windows hasn't uploaded yet due to the issue noted below ( https://github.com/conda-forge/cudatoolkit-feedstock/pull/57#issuecomment-811311604 ), but should be uploaded soon as well. Other versions haven't seen an upload in over a month

Screen Shot 2021-03-31 at 11 33 00 AM

Comment (isuruf) [reply recipe/conda_build_config.yaml:9]: > Other versions haven't seen an upload in over a month

That's because the gcc compiler version hasn't changed in over a month.

Comment (jakirkham) [reply recipe/conda_build_config.yaml:9]: If we are concerned about this, we could add the compilers to `ignore_version`

https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#special-variant-keys
https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#avoiding-unnecessary-builds

**Comment by jakirkham** [comment]
Windows CUDA 11.2.2 build completed. Packages have been uploaded, but will take ~30mins to mirror to CDN

**Comment by jakirkham** [comment]
`conda-forge-ci-setup` has also been rebuilt and uploaded. Mirroring to CDN atm, which should complete in ~30mins

**Pull request merged**
